### DISCUSSION TITLE
Problem in Funktion weewx_wdc.py/get_calendar_data

### DISCUSSION THREAD

**Discussion by hoetzgit** [Daveiano/weewx-wdc#76] [Q&A] [answered]
Hallo David,

aus bestimmten Gründen habe ich in einigen WeeWX Datenbanktabellen Zeilen mit dateTime belegt, aber den Rest der Spalten mit NULL Werten. Das ist beabsichtigt und macht bei mir für bestimmte Situationen Sinn. Es führt jedoch bei weewx-wdc zu folgendem Problem (siehe unten). Da ich von Python nicht so viel Ahnung habe, schreibe ich hier wie ich es gelöst habe und würde Dich bitten, Dir das mal anzusehen und zu bewerten, ggf. in irgendeiner Form bei Dir einzubauen.

Vielen Dank und Grüße Henry

Fehlermeldung:

```
Oct 14 08:55:47 weewx weewx[303151] ERROR weewx.cheetahgenerator: Evaluation of template /home/weewx/skins/weewx-wdc/statistics.html.tmpl failed with exception '<class 'TypeError'>'
Oct 14 08:55:47 weewx weewx[303151] ERROR weewx.cheetahgenerator: **** Ignoring template /home/weewx/skins/weewx-wdc/statistics.html.tmpl
Oct 14 08:55:47 weewx weewx[303151] ERROR weewx.cheetahgenerator: **** Reason: '>' not supported between instances of 'NoneType' and 'float'
Oct 14 08:55:47 weewx weewx[303151] ERROR weewx.cheetahgenerator: ****  Traceback (most recent call last):
Oct 14 08:55:47 weewx weewx[303151] ERROR weewx.cheetahgenerator: ****    File "/home/weewx/bin/weewx/cheetahgenerator.py", line 348, in generate
Oct 14 08:55:47 weewx weewx[303151] ERROR weewx.cheetahgenerator: ****      unicode_string = compiled_template.respond()
Oct 14 08:55:47 weewx weewx[303151] ERROR weewx.cheetahgenerator: ****    File "_home_weewx_skins_weewx_wdc_statistics_html_tmpl.py", line 192, in respond
Oct 14 08:55:47 weewx weewx[303151] ERROR weewx.cheetahgenerator: ****    File "/usr/lib/python3/dist-packages/Cheetah/Template.py", line 1708, in _handleCheetahInclude
Oct 14 08:55:47 weewx weewx[303151] ERROR weewx.cheetahgenerator: ****      self._CHEETAH__cheetahIncludes[_includeID].respond(trans)
Oct 14 08:55:47 weewx weewx[303151] ERROR weewx.cheetahgenerator: ****    File "_home_weewx_skins_weewx_wdc_includes_climatological_days_inc.py", line 330, in respond
Oct 14 08:55:47 weewx weewx[303151] ERROR weewx.cheetahgenerator: ****    File "_home_weewx_skins_weewx_wdc_includes_climatological_days_inc.py", line 127, in __errorCatcher9
Oct 14 08:55:47 weewx weewx[303151] ERROR weewx.cheetahgenerator: ****    File "<string>", line 1, in <module>
Oct 14 08:55:47 weewx weewx[303151] ERROR weewx.cheetahgenerator: ****    File "/home/weewx/bin/user/weewx_wdc.py", line 1030, in get_calendar_data
Oct 14 08:55:47 weewx weewx[303151] ERROR weewx.cheetahgenerator: ****      for day in days:
Oct 14 08:55:47 weewx weewx[303151] ERROR weewx.cheetahgenerator: ****    File "/home/weewx/bin/user/weewx_wdc.py", line 1026, in <lambda>
Oct 14 08:55:47 weewx weewx[303151] ERROR weewx.cheetahgenerator: ****      lambda x: x[1].raw > 0.0, list(zip(day_series.start, day_series.data))
Oct 14 08:55:47 weewx weewx[303151] ERROR weewx.cheetahgenerator: ****  TypeError: '>' not supported between instances of 'NoneType' and 'float'
```

Korrektur (Filter angepasst)

```
    def get_calendar_data(self, obs, aggrgate_type, period):
        """
        Returns array of calendar data for use in diagram.

        Args:
            observation (string): The observation
            aggrgate_type (string): Min, max, avg.
            period (obj): Period to use, eg. $year, month, $span

        Returns:
            list: Calendar data.
        """
        if obs == "rain":
            day_series = period.rain.series(
                aggregate_type=aggrgate_type,
                aggregate_interval="day",
                time_series="start",
                time_unit="unix_epoch",
            ).round(self.diagram_util.get_rounding("rain"))

            days = filter(
                lambda x: (x[1].raw > 0.0) if x[1].raw is not None else False, list(zip(day_series.start, day_series.data))
            )
            rainDays = []

            for day in days:
                rainDays.append({"value": day[1].raw, "day": day[0].format("%Y-%m-%d")})

            return rainDays

        if obs == "outTemp":
            day_series = period.outTemp.series(
                aggregate_type=aggrgate_type,
                aggregate_interval="day",
                time_series="start",
                time_unit="unix_epoch",
            ).round(self.diagram_util.get_rounding("outTemp"))

            days = list(zip(day_series.start, day_series.data))
            tempDays = []

            for day in days:
                if day[1].raw is not None:
                    tempDays.append(
                        {"value": day[1].raw, "day": day[0].format("%Y-%m-%d")}
                    )
            return tempDays
```

**Comment by Daveiano** [comment]
Hallo Henry, 

Danke für deinen Beitrag! Das ist eine absolut sinnvolle Änderung, ~~NULL Werte können auch zustande kommen wenn z.b die Station über mehrere Tage keine Daten sendet (batterien leer)~~. Habe grade getestet und gemerkt das weewx keine NULL werte für "leere" tage in die Datenbank speichert, die Tage sind dann einfach nicht vorhanden.

Ich schaue es mir an und werde es in die bald erscheinende Version 2.3.0 mit einbauen!

**Comment by hoetzgit** [comment, marked as answer]
Hallo David, ja, deshalb habe ich auch keinen normalen Fehler gemeldet, da der "Fehler" auf meine Datenmanipulation zurückzuführen ist. 

Ich benutze als Hauptseite die Belchertown Skin. Start meiner Station war der August 2019. Um aber Graphen für das Jahr 2019 komplett darzustellen, habe ich in die `archiv` Tabelle jeweils pro Monat Jan.-Jul. 2019 einen dummy Datensatz mit `NULL` Werten eingefügt. Da im Laufe der Zeit neue Sensoren und Erweiterungen (dadurch neue Werte und DB Felder) hinzugekommen sind, habe ich zur Vervollständigung der DB Werte `wee_database` mit `--calc-missing` und anschließend mit `--rebuild_daily` laufen lassen. Dadurch werden mir aber in den `archiv_day_xxx` Tabellen in den Datensätzen vor dem jeweilen Start der Messungen/Berechnungen diese -im Prinzip ja richtigen - `sum=0.0, count=0, wsum=0.0, sumtime=0` Werte erzeugt. Das führte dazu, dass in Tabellen auf der Webseite z.B. "0.0" angezeigt wird. Das wollte ich aber nicht, da die Sensoren in dieser Zeit noch nicht installiert waren und "0.0" dann irgendwie nicht richtig ist. Also habe ich auch diese Werte mit `dateTime` vor dem jeweiligen Start auf `NULL` gesetzt. Webseiten-Tabellen zeigen jetzt für diese Zeiträume "-" statt "0.0" an. 

Lange Rede kurzer Sinn, ich wusste mir nur durch diese Maßnahmen zu helfen und dadurch trat der "Fehler" überhaupt erst auf.

Aber so konnte ich wenigstens in Deiner Anpassungen sehen, wie man das richtig programmiert und habe wieder etwas gelernt.

Nochmals Danke und viele Grüße
Henry

**Comment by Daveiano** [comment]
Danke für die ausführliche Beschreibung, ich verstehe!

Ja wie du es gemacht hast geht es auch, ist nur etwas schwerer lesbar finde ich, aber es erfüllt seinen Zweck.

Danke zurück und liebe Grüße

**Comment by hoetzgit** [comment]
Hallo David,

mir fällt gerade etwas zu 2.3.0 ein. Habe die Version heute Vormittag
installiert und ich glaube mich zu erinnern, dass kurz ein Fehler
durchgehuscht ist, sinngemäss show_yesterday not defined. Kann es aber
gerade nicht prüfen, da ich unterwegs bin. Evtl. schaust Du mal, ob das
noch etwas fehlt.

Grüsse
Henry

Am 16. Oktober 2022 01:17:34 schrieb Daveiano ***@***.***>:

**Reply by Daveiano** [reply]
Hallo Henry,

habe nochmal getestet und bei mir sind keine Fehler aufgetreten, falls der Fehler nochmal auftritt oder du ihn nochmal in den logs findest, lege gerne ein issue dafür an. 

Deutet aber natürlich auf ein eventuelles Problem mit der neuen "Gestern-Seite" hin, mehr dazu hier: https://github.com/Daveiano/weewx-wdc/wiki/Configuration#yesterday-page

**Reply by hoetzgit** [reply]
Hallo David, ich kann tatsächlich diesen Fehler nicht mehr reproduzieren. Was ja eigentlich gut ist, da ich wahrscheinlich ein Konfigurationsproblem hatte.

Edit: DWD: Hast Du evtl. "SchilderLZ.jpg" statisch programmiert?

Viele Grüße
Henry
